### PR TITLE
chore(clerk-js): React18 for clerk-js and shared

### PR DESCRIPTION
**Pull request by dvargas92495**
## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [x] 📖 Refactoring / dependency upgrade / documentation
- [x] other:

## Packages affected

- [x] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/clerk-expo`
- [x] `@clerk/backend-core`
- [ ] `@clerk/clerk-sdk-node`
- [x] `@clerk/edge`
- [x] `build/tooling/chore`

## Description

- [ ] `npm test` runs as expected.
    - Issues listed below
- [x] `npm run build` runs as expected.

I started work on https://github.com/clerkinc/javascript/issues/220 - ran into a few issues running clerk build/tests locally so I wanted to make a PR for it to prompt further discussion.

Issues:
- Engine requirements - I'm currently have to run Node14 locally because [AWS Lambda doesn't yet support 16](https://github.com/aws/aws-lambda-base-images/issues/14#issuecomment-1108679026), though they plan to later this month. However, if there's nothing specifically preventing being run on node14, might as well drop the requirement to 14 or drop it all together.
- `TURBO_CONCURRENCY` - I did not have this env defined. It should either be defined for the developer by default or listed in the setup instructions
- `./moduleTypeFix` - converted these files to `node` scripts since windows doesn't support running `./` commands. but everyone running this package will have `node`
- React test failures - This is the biggest issue I need help on - A ton of tests are failing for me locally due to `Cannot read property 'useState' of null`; basically `React` is null throughout the tests. I'm not sure exactly why this is or if this is related to bumping up to react 18 yet. React testing library does not yet support React 18 from what I've read, but I don't think that should imply `React` resolving to `null`.

After fixing the above, `clerk-js` was able to build correctly on React 18. Testing on React 18 is blocked by the test failures I mentioned above

### CODE REVIEW

File: packages/clerk-js/src/ui/common/withRedirectToHome.tsx [hunk 1]
```diff
@@ -16,7 +16,7 @@ export function withRedirectToHome<P extends SignInProps | SignUpProps>(
     const { singleSessionMode } = authConfig;
     const session = useCoreSession({ avoidUndefinedCheck: true });
 
-    React.useEffect(() => {
+    useEffect(() => {
```
Comment (dvargas92495) [inline packages/clerk-js/src/ui/common/withRedirectToHome.tsx:19]: This is an example of a file throwing an error in tests. Swapping the imports didn't help

File: packages/clerk-js/src/ui/styles/_userprofile.scss [hunk 2]
```diff
@@ -67,7 +67,7 @@ $content-width-lg: 620px;
       border-right-color: transparent;
       width: 26px;
       height: 26px;
-      border-radius: (26px / 2);
+      border-radius: 50%;
```
Comment (dvargas92495) [inline packages/clerk-js/src/ui/styles/_userprofile.scss:70]: This was throwing a warning for not using `calc`. Based on context, seems like border radius `50%` will do the same thing for cheaper

File: packages/shared/utils/svgTransform.js [hunk 3]
```diff
@@ -9,7 +9,7 @@ module.exports = {
     const code = babel.transform(
       `
         import React from 'react';
-        export default () => (<svg data-filename="${path.relative(process.cwd(), filename)}" />);
+        export default () => (<svg data-filename="${path.relative(process.cwd(), filename).replace(/\\/g, '/')}" />);
```
Comment (dvargas92495) [inline packages/shared/utils/svgTransform.js:12]: This was another source of failure for tests, since windows uses `\\` file path delimiters. Though, this didn't seem to make a difference. I thought based on the jest config this was where `svg`s were being transformed for tests, is there somewhere else to look at?

**Comment by dvargas92495** [comment]
tagging @nikosdouvlis for preemptive review and question surrounding React being null

**Comment by dvargas92495** [comment]
Ah reading the CI failure, I'm seeing "Invalid hook call" which usually happens with multiple react versions. I think I know where to debug from here

File: packages/clerk-js/src/ui/common/withRedirectToHome.tsx (hunk 1 above)
Comment (SokratisVidros) [reply packages/clerk-js/src/ui/common/withRedirectToHome.tsx:19]: @dvargas92495 I am not sure I get what the problem was with tests and this change. Can you please elaborate?

Comment (dvargas92495) [reply packages/clerk-js/src/ui/common/withRedirectToHome.tsx:19]: looking at the CI failure, I'm now pretty sure it's due to the mismatched React versions between tests and now clerk-js using React 18:

image

As I mentioned in the comment below, I think I know how to iterate from here and am unblocked for now

File: packages/clerk-js/package.json [hunk 4]
```diff
@@ -46,8 +46,8 @@
     "deepmerge": "^4.2.2",
     "js-cookie": "^2.2.1",
     "qs": "^6.9.4",
-    "react": "17.0.2",
-    "react-dom": "17.0.2",
+    "react": "^18.1.0",
```
[the remark this answers is not in the document]
Comment (SokratisVidros) [reply packages/clerk-js/package.json:49]: You can do an `npm update -w workspace` for each workspace and then a global npm install at the root folder.

File: package.json [hunk 5]
```diff
@@ -10,7 +10,7 @@
   "private": true,
   "packageManager": "npm@8.5.0",
   "engines": {
-    "node": ">=16.8.0",
+    "node": ">=14",
```
Comment (SokratisVidros) [inline package.json:13]: @dvargas92495 can you please explain why is this necessary?

File: packages/clerk-js/src/ui/common/withRedirectToHome.tsx (hunk 1 above)
Comment (SokratisVidros) [reply packages/clerk-js/src/ui/common/withRedirectToHome.tsx:19]: Yes, that should be it. Let's make sure that only React 18 is used.

File: packages/shared/package.json [hunk 6]
```diff
@@ -16,9 +16,9 @@
     "@svgr/webpack": "^6.2.1",
     "core-js": "^3.18.3",
     "js-cookie": "^2.2.1",
-    "react": "17.0.2",
+    "react": ">=16",
```
Comment (SokratisVidros) [inline packages/shared/package.json:19]: ❓ Shouldn't we use React 18?

File: packages/shared/utils/svgTransform.js (hunk 3 above)
Comment (SokratisVidros) [reply packages/shared/utils/svgTransform.js:12]: Did the snapshot tests fail due to a missmatch in the SVG delimiter in windows?

Comment (dvargas92495) [reply packages/shared/utils/svgTransform.js:12]: yup

File: packages/clerk-js/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (dvargas92495) [reply packages/clerk-js/package.json:49]: Cool, going to try this locally, thanks!

**Comment by SokratisVidros** [comment]
@dvargas92495 some tests are still failing. Could you please have another look?

File: package.json (hunk 5 above)
Comment (dvargas92495) [reply package.json:13]: https://aws.amazon.com/about-aws/whats-new/2022/05/aws-lambda-adds-support-node-js-16/

Now I'm definitely going to revert this back, I can finally move to node 16 locally!

**Review by dvargas92495 — COMMENTED** [review]
Ok all tests are passing, I think the PR is officially ready for review! Taking out of draft

It's a big one, so I'm happy to split this out into smaller PRs if that would make things easier to merge. Package-lock in particular is scary so I would first do all the test changes as one PR, then the package upgrades as the second.

ccing: @nikosdouvlis + @SokratisVidros for review

I know according to https://github.com/clerkinc/javascript/issues/220, we also wanted to do some manual testing - how do you guys typically inspect package changes in the browser locally? There's no Storybook so figured there might be something in the repo that helps spin up the changes for development

File: package.json (hunk 5 above)
Comment (SokratisVidros) [reply package.json]: Right on time!

**Comment by SokratisVidros** [comment]
> Ok all tests are passing, I think the PR is officially ready for review! Taking out of draft
> 
> It's a big one, so I'm happy to split this out into smaller PRs if that would make things easier to merge. Package-lock in particular is scary so I would first do all the test changes as one PR, then the package upgrades as the second.
> 
> ccing: @nikosdouvlis + @SokratisVidros for review
> 
> I know according to #220, we also wanted to do some manual testing - how do you guys typically inspect package changes in the browser locally? There's no Storybook so figured there might be something in the repo that helps spin up the changes for development

Thanks @dvargas92495. We have tests applications locally for all ClerkJS and Clerk React components. These applications will be publicly available soon.

**Comment by SokratisVidros** [comment]
@dvargas92495 Did you notice an increased duration in tests due to the addition of all the `await` before each user interaction? What's the official recommendation by https://testing-library.com/ for React 18?

**Comment by dvargas92495** [comment]
Just rebased to fix the new failing test

File: packages/clerk-js/package.json [hunk 7]
```diff
@@ -113,7 +113,7 @@
     "files": [
       {
         "path": "./dist/clerk.js",
-        "maxSize": "162kB"
+        "maxSize": "164kB"
```
Comment (dvargas92495) [inline packages/clerk-js/package.json:116]: Not sure what the policy on bundle watch here is, but it went up on the latest rebase

**Review by dvargas92495 — COMMENTED** [review]
At the risk of another rebase, any further thoughts on this PR @nikosdouvlis @SokratisVidros on next move? My inclination is that is should be safe to merge, but haven't done any manual UI testing yet

**Review by SokratisVidros — APPROVED** [review]

**Review by nikosdouvlis — APPROVED** [review]
Hey @dvargas92495 - this is looking really good :) 
We will release the latest SignUp improvements and we'll merge it right after, so please expect it to be merged either today or tomorrow.

Thanks!
